fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,273 @@
+    poll.start();
+
+    Ok(())
+}
+
+impl<T> PollingParentSyncer<T> {
+    pub fn new(
+        config: Config,
+        parent_view_provider: Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
+        agent: Arc<IPCProviderProxy>,
+        query: Arc<T>,
+    ) -> Self {
+        Self {
+            config,
+            parent_view_provider,
+            agent,
+            committed_state_query: query,
+        }
+    }
+}
+
+impl<T: ParentFinalityStateQuery + Send + Sync + 'static> PollingParentSyncer<T> {
+    /// Start the parent finality listener in the background
+    pub fn start(self) {
+        let config = self.config;
+        let provider = self.parent_view_provider;
+        let agent = self.agent;
+        let query = self.committed_state_query;
+
+        let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+        tokio::spawn(async move {
+            loop {
+                interval.tick().await;
+
+                if let Err(e) = sync_with_parent(&config, &agent, &provider, &query).await {
+                    tracing::error!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+    }
+}
+
+/// Syncing with parent with the below steps:
+/// 1. Get the latest height in cache or latest height committed increment by 1 as the
+///    starting height
+/// 2. Get the latest chain head height deduct away N blocks as the ending height
+/// 3. Fetches the data between starting and ending height
+/// 4. Update the data into cache
+async fn sync_with_parent<T: ParentFinalityStateQuery + Send + Sync + 'static>(
+    config: &Config,
+    agent_proxy: &Arc<IPCProviderProxy>,
+    provider: &Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
+    query: &Arc<T>,
+) -> anyhow::Result<()> {
+    let last_recorded_height = if let Some(h) = last_recorded_height(provider).await? {
+        h
+    } else {
+        // cannot get starting recorded height, we just wait for the next loop execution
+        return Ok(());
+    };
+
+    let parent_chain_head_height = agent_proxy
+        .get_chain_head_height()
+        .await
+        .context("cannot fetch parent chain head")?;
+    // sanity check
+    if parent_chain_head_height < config.chain_head_delay {
+        tracing::debug!("latest height not more than the chain head delay");
+        return Ok(());
+    }
+
+    let ending_height = parent_chain_head_height - config.chain_head_delay;
+
+    tracing::debug!(
+        "last recorded height: {}, parent chain head: {}, ending_height: {}",
+        last_recorded_height,
+        parent_chain_head_height,
+        ending_height
+    );
+
+    if last_recorded_height == ending_height {
+        tracing::debug!(
+            "the parent has yet to produce a new block, stops at height: {last_recorded_height}"
+        );
+        return Ok(());
+    }
+
+    // we are going backwards in terms of block height, the latest block height is lower
```

Comment:
I guess for now this is good enough, it may have a bit of bad performance but I am personally fine with it.